### ISSUE TITLE
Front-end: Can vertical ordering of variants in beadplot be optimized?

### ISSUE THREAD

**Issue by ArtPoon** [PoonLab/covizu#58]
Presently, the length of a vertical edge bears no relation to the genetic distance (*i.e.*, number of mutations) it represents.  Variants are ordered by pre-order traversal, which helps keep subtrees together.  

Is there a more optimal arrangement of variants possible?

**Comment by ArtPoon** [comment]
The beadplot we use for the README document actually provides a useful example:
![](https://github.com/PoonLab/covizu/blob/master/doc/beadplot.png?raw=true)

The two beads at the bottom of the plot the following distances to the "baseline" variant at the top of the plot (Scotland/EDB058):
* Wales/PHWC-2B746 - 1 mutation
* Scotland/EDB2095 - 1 mutation

but the variant Netherlands/NA_17 is near the top of the plot and has a distance of 4.02 mutations.
The cluster of samples from Australia (rooted by Australia/VIC774) is located "above" the Wales and Scotland variants and has a distance of 4 mutations.

There are total of 9 child variants that descend from Scotland/EDB058.

**Comment by ArtPoon** [comment]
Now sorting vertical position of variant by number of mutations from parental variant, using collection date of earliest sample of variants to break ties

**Comment by ArtPoon** [comment]
New idea (the above isn't bad, but this could be better) - instead of ordering children by genetic distance and then collection date for tie-breaking, what about genetic distance and then total number of descendant variants?  The problem is that some variants that are only 1 mutation away are being bounced downwards by an earlier variant that is ancestral to a large number of descendant variants.

**Comment by SyouTono242** [comment]
Sorting with ascending recursive width (last date sampled for all children - first date sample for all children) and then initial sample date for all subtrees. 
It seems to be working now but the unsampled ones are sticking out for some reason. I think it has something to do with NA being added to the total length... I'll be trying to assign some value (e.g. max date on the graph) to see if we can solve the issue.
<img width="784" alt="image" src="https://user-images.githubusercontent.com/56494700/176245578-80887b7b-2014-443a-90e0-3b69c2aa42b7.png">

**Comment by ArtPoon** [comment]
Unsampled lineages can inherit the earliest sample date from their "parent" lineage (if that is sampled) - if parent is also unsampled, then proceed down the tree until you reach a sampled parent.

**Comment by ArtPoon** [comment]
We should consider re-rooting these neighbor-joining trees so that the sampled lineage with the earliest sample date (and also largest number of descendants as a tie-breaker?) becomes the root.

**Comment by ArtPoon** [comment]
@SyouTono242 can you please start commiting your code changes to a branch of this repo? thanks!

**Comment by SyouTono242** [comment]
@ArtPoon Yes... But I was an idiot who forgot to checkout and I accidentally committed **and** pushed my script to **master**... I am extremely sorry; I shouldn't be checking GitHub at 4am and I learnt my lesson hard... I didn't commit anything else than the one single script I wrote, and as it's a standalone script it shouldn't interfere with any other existing file. Please roll back masters if you can... Again I apologize for all the inconvenience if anyone else is fetching the incorrect changes. I'm really sorry.

**Comment by ArtPoon** [comment]
Haha whoops! Don’t stress about it @SyouTono242 - accidents happen and I’ve made the same mistake before.  @GopiGugan can you please roll back this commit and @SyouTono242 can you push your changes to the dev branch?

**Comment by ArtPoon** [comment]
Code commited to `dev` branch, incorporate into a PR and review affects of new code on beadplots

**Comment by ArtPoon** [comment]
Keep an eye on this new code while reviewing PR #409

**Comment by ArtPoon** [comment]
Let's work on porting this into JS for the next PR (after #409)

**Comment by ArtPoon** [comment]
@GopiGugan to generate toy data sets to run `beadplot.py` script
@bonnielu to port @SyouTono242 's code from R to Python (in `beadplot.py`)

**Comment by ArtPoon** [comment]
Apply sorting algorithm directly to data structure before it is serialized as a JSON file

**Comment by ArtPoon** [comment]
@bonnielu has test data from @GopiGugan , code analysis in progress

**Comment by ArtPoon** [comment]
porting is finished, @bonnielu requesting larger test data

**Comment by ArtPoon** [comment]
We need to write some unit test fixtures (trees that should be re-ordered a specific way) to determine why these test data sets are not being changed.

**Comment by bonnielu** [comment]
I tested it out using the data that Gopi gave me and this is what the beadplots look like after reordering by date! The larger lineages do seem like they have a tendency to line up if we reorder by this method. 

![image](https://user-images.githubusercontent.com/42049026/193832908-e04db793-fd22-47fc-ad59-2af87282d073.png)
![image](https://user-images.githubusercontent.com/42049026/193833092-79bd8d33-7ff3-441d-83bf-7c130eff8ba7.png)

**Comment by GopiGugan** [comment]
![Screen Shot 2022-10-04 at 1 54 35 PM](https://user-images.githubusercontent.com/35086217/193891383-be86ae08-4350-4e6a-8376-1b547a97549d.png)
![Screen Shot 2022-10-04 at 1 54 57 PM](https://user-images.githubusercontent.com/35086217/193891400-1900d324-b2b1-4583-b6cd-a11f81038d1f.png)

**Comment by ArtPoon** [comment]
Ok thanks!  I'm willing to call the first, smaller beadplot an improved layout.  Difficult to say for the second one - but I wonder what the heck is going on with all those samples taken on the same date?

**Comment by ArtPoon** [comment]
Can we please see what the beadplots look like if we prioritize earliest sampling date, and then tree traversal?

**Comment by bonnielu** [comment]
This is what it looks like if we prioritize sorting by (last date - first date), earliest sampling date, followed by level-order tree traversal. 

![Screen Shot 2022-10-11 at 11 03 42 AM](https://user-images.githubusercontent.com/42049026/195147704-d864799e-a2b6-4d1c-9723-860a50a34e48.png) ![Screen Shot 2022-10-11 at 11 03 28 AM](https://user-images.githubusercontent.com/42049026/195147740-200d3215-1f32-40c6-a151-a3a188125d67.png)

**Comment by ArtPoon** [comment]
Thanks for implementing this different approach @bonnielu - after reviewing the outputs, I think we should go ahead with the first revision (Yiran's version).  Please revert to that algorithm and submit a PR to dev.

**Comment by ArtPoon** [comment]
I don't think we're going to find a good solution to this - the problem might be better addressed by collapsing variants (#434)
